Make name content attributes consistently atoms and put them in rare_data for fast access

pshaughn opened the pull request:
<!-- Please describe your changes on the following line: -->
All codepaths setting the name content attribute now use an atom, which is also stored in rare_data for direct lookup by a get_name method. 

Paralleling the get_name method, I added a get_id method, which makes some internal id-lookup cases nicer.

A new test tests for a name setter on every HTML element type. In addition to its overt and upstreamable purpose of checking IDL property reflection semantics, for us this test also hits some Servo assertions that make sure the name is an atom in every case. If the test doesn't crash, even a failed test case still has the attribute as an atom rather than some other type. The failed cases are for elements that we have unimplemented or completely stubbed; I added a few missing name IDL properties to otherwise implemented elements.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #25570 and make progress on #25057

<!-- Either: -->
- [X] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

pshaughn committed:
Names should now be consistently atoms

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/htmltextareaelement.rs, components/script/dom/htmlcollection.rs, components/script/dom/htmlmetaelement.rs, components/script/dom/htmlselectelement.rs, components/script/dom/webidls/HTMLOutputElement.webidl and 18 more
 * @kichjang: components/script/dom/htmltextareaelement.rs, components/script/dom/htmlcollection.rs, components/script/dom/htmlmetaelement.rs, components/script/dom/htmlselectelement.rs, components/script/dom/webidls/HTMLOutputElement.webidl and 18 more

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/web-platform-tests/wpt/pull/21332.

pshaughn commented:
I am sure this will have some idlharness test result changes, and I wouldn't be surprised if filling in the name setter has had other effects.
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 8112cea249182d72ec28ee2c7cd7ffb3b2393293 with merge 513b937cbd8bf368cff7a55e1594375275837f6b...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/IXngXKBvRiGDE06lpBUBmg)

pshaughn commented:
That one /css/css-transforms/animation/perspective-interpolation.html crash looks scary. Other new failures are probably because we were only "accidentally" passing before, but of course I'll have to look through them all to be sure.

pshaughn commented:
Oh, I was reading upwards from the bottom and missed the known-intermittents header, that one test already has an intermittency issue filed for it. Now to look at all the real new failures...

pshaughn commented:
Once I was reading the right part of the results, it was all new passes and completely consistent with what I changed. I notice the CI test harness does a better job of extracting results from a timed-out test than I get running on the command line; the _mozilla test I added is not actually necessary for CI purposes.

pshaughn commented:
I believe this is ready to roll out.
@bors-servo try

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#21332.

bors-servo commented:
:hourglass: Trying commit b721b6cf41f8995a37cdea70dbf75ca737af1679 with merge 040cf69bc6ec3c5f75fb3b5666b2b2286f9c081c...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/c-EEP2e2TrKHkdAVWiQR5w)

jdm commented:
That's #23290.

jdm commented on components/script/dom/element.rs:
```diff
@@ -1388,11 +1388,27 @@ impl Element {
             .find(|a| a.name() == name)
-            .map(|js| DomRoot::from_ref(&**js))
+            .map(|js| DomRoot::from_ref(&**js));
+        fn id_and_name_must_be_atoms(name: &LocalName, maybe_attr: &Option<DomRoot<Attr>>) -> bool {
```
We probably need a `#[cfg(debug_assertions)]` to avoid warnings about an unused function in builds without them enabled.

jdm reviewed (approved):
Thanks for doing this work! I'm impressed by how many places ended up needing to be modified.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#21332.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 2891c8b has been approved by `jdm`

<!-- @bors-servo r=jdm 2891c8bba0f3153f3d575e0bb11bbca0fbd8d7b4 -->

bors-servo commented:
:hourglass: Testing commit 2891c8bba0f3153f3d575e0bb11bbca0fbd8d7b4 with merge 5d152afbf5b3b0e1193148f9a946c579fb0f927c...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/QMkjWsQzSFeHnkawzZQMkQ)

jdm commented:
Sorry, I was wrong about how the debug_assert macro is treated by the compiler:
```
error[E0425]: cannot find function `id_and_name_must_be_atoms` in this scope
    --> components/script/dom/element.rs:1425:23
     |
1425 |         debug_assert!(id_and_name_must_be_atoms(name, &maybe_attribute));
     |                       ^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope
```

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#21332.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit fe16a65 has been approved by `jdm`

<!-- @bors-servo r=jdm fe16a659f5f810ba9ed954a78f6598132ccec3c8 -->

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#21332.

jdm commented:
This needs a `mach fmt`.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#21332.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 02a3e59 has been approved by `jdm`

<!-- @bors-servo r=jdm 02a3e59fd33f0f22cba46d32ecfd5e98481cf095 -->

bors-servo commented:
:hourglass: Testing commit 02a3e59fd33f0f22cba46d32ecfd5e98481cf095 with merge 2b2b3ea0db79fe1ea042019aa3163f59db2b5aa3...

pshaughn commented:
The spec sometimes explicitly excludes the empty-string from being a possible name and sometimes doesn't; there aren't WPT tests on the subject so I don't know what browsers do in various name="" cases, and empty-string handling here might not be great.

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/W-Lr6kxsTKCMXu_7bqpxGg)
Approved by: jdm
Pushing 2b2b3ea0db79fe1ea042019aa3163f59db2b5aa3 to master...

bors-servo commented:
:eyes: Test was successful, but fast-forwarding failed: 422 Update is not a fast forward

pshaughn commented:
I don't know what that means but it sounds like something a rebase might fix.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#21332.

jdm commented:
@bors-servo r+
Yeah, those failures are mysterious :(

bors-servo commented:
:pushpin: Commit f29e22f has been approved by `jdm`

<!-- @bors-servo r=jdm f29e22f131291ed1bcd581cb9be6807c66c1534e -->

bors-servo commented:
:hourglass: Testing commit f29e22f131291ed1bcd581cb9be6807c66c1534e with merge 9c135f0e9554bd78bb7a432a872257e2cae13a48...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/QDhI56eBQOixBw4y06yAZw)
Approved by: jdm
Pushing 9c135f0e9554bd78bb7a432a872257e2cae13a48 to master...
